### PR TITLE
OpenBSD build fix to silent not supported features mainly.

### PR DESCRIPTION
**Pull request by devnexen**

### CODE REVIEW

**Comment by vuvova** [comment]
What's wrong with RocksDB and TokuDB on OpenBSD?

**Comment by devnexen** [comment]
Nothing "wrong" as such, just OpenBSD does not have Direct I/O support and no timer functions support.

**Comment by vuvova** [comment]
What happens with rocksdb and tokudb without your patch? fails to compile? how?

**Comment by devnexen** [comment]
Yes exactly compilation failure since tokudb is known for high performance, so even if we disable the direct I/o part in the code it is pointless in my opinion. For rocksdb, same compilation failure too because no timer_* functions.

**Comment by vuvova** [comment]
I see. Then, I think, it's tokudb's and rocksdb's bug that they don't properly check for prerequisites.
Options:
1.  you put proper checks in corresponding CMakeLists.txt files. Not for OpenBSD, but for required prerequisites
2. You simply paste here the compilation errors for rocksdb and tokudb and I'll do the checks :)

**Comment by devnexen** [comment]
Ok fair enough I ve changed let me know if the changes look better to you.

File: storage/tokudb/PerconaFT/CMakeLists.txt [hunk 1]
```diff
@@ -52,6 +52,10 @@ include(TokuThirdParty)
 set(TOKU_CMAKE_SCRIPT_DIR "${CMAKE_CURRENT_SOURCE_DIR}/cmake")
 include(TokuMergeLibs)
 
+if (NOT HAVE_O_DIRECT AND NOT HAVE_F_NOCACHE)
+   RETURN()
+endif()
+
```
Comment (vuvova) [inline storage/tokudb/PerconaFT/CMakeLists.txt]: This wouldn't do. Try to build. You disable PerconaFT library, but not the tokudb plugin itself. It'll fail to build because some of its dependencies won't be built.

The check is fine, but you need it in storage/tokudb/CMakeLists.txt

**Comment by devnexen** [comment]
Good points. I changed accordingly but keeping the modules into PerconaFT I guess if really needed it one day will be moved in the upper level

File: storage/connect/reldef.cpp [hunk 2]
```diff
@@ -23,6 +23,9 @@
 #include <dlfcn.h>          // dlopen(), dlclose(), dlsym() ...
 #include "osutil.h"
 //#include "sqlext.h"
+#if defined(__OpenBSD__)
+#define RTLD_NOLOAD 0
+#endif
 #endif
```
Comment (vuvova) [inline storage/connect/reldef.cpp]: ok, got  it. although, I'd rather remove RTLD_NOLOAD from the code below, I don't see any need for it.

File: storage/mroonga/vendor/groonga/CMakeLists.txt [hunk 3]
```diff
@@ -242,6 +242,7 @@ ac_check_lib(dl dlopen)
 ac_check_lib(execinfo backtrace)
 if(HAVE_LIBEXECINFO)
   set(HAVE_BACKTRACE TRUE)
+  set(HAVE_EXECINFO_H TRUE)
 else()
```
Comment (vuvova) [inline storage/mroonga/vendor/groonga/CMakeLists.txt]: Why is that needed? HAVE_EXECINFO_H is checked and set in `configure.cmake`:
```
CHECK_INCLUDE_FILES (execinfo.h HAVE_EXECINFO_H)
```

File: storage/tokudb/CMakeLists.txt [hunk 4]
```diff
@@ -16,6 +16,14 @@ IF(NOT TOKUDB_OK)
   RETURN()
 ENDIF()
 
+set(CMAKE_MODULE_PATH "${CMAKE_CURRENT_SOURCE_DIR}/PerconaFT/cmake_modules")
+include(TokuFeatureDetection)
+
```
Comment (vuvova) [inline storage/tokudb/CMakeLists.txt]: I don't quite like that you pull in TokuFeatureDetection from PerconaFT here. And that it performs lots of checks even of tokudb is disabled.

Instead, I'd extend the TOKUDB_OK check to test for F_NOCACHE and O_DIRECT.

File: storage/connect/reldef.cpp (hunk 2 above)
Comment (devnexen) [reply storage/connect/reldef.cpp]: It is needed because OpenBSD does not have this flag.

File: storage/mroonga/vendor/groonga/CMakeLists.txt (hunk 3 above)
Comment (devnexen) [reply storage/mroonga/vendor/groonga/CMakeLists.txt]: Yes I knew but it did not work. Usually find_path works better (backtrace needs an external library and is not in system folders neither in BSD unlike Linux). But I m not against removing it.

File: storage/tokudb/CMakeLists.txt (hunk 4 above)
Comment (devnexen) [reply storage/tokudb/CMakeLists.txt]: Oh ok yes good point.

**Comment by devnexen** [comment]
Feel free to do the changes but in mind with my comments. Thanks.

**Comment by vuvova** [comment]
If you'd like you can check the referenced commit (above) to see whether it solves the issue.

**Comment by devnexen** [comment]
Looks good will check later but I m pretty confident at the moment. Thanks for your time.

**Comment by devnexen** [comment]
I had one question, if all good is it not too much to backport it to 10.1 for example ?

**Comment by vuvova** [comment]
Sure. Did you verify that it helps?

**Comment by devnexen** [comment]
in a couple of hours I will good it s possible to backport then :-)

**Comment by devnexen** [comment]
As expected all good, the only failure is not related to the server code but the c connector (which I did a PR for) so you can commit :-) and nicely back port it to 10.1 Thanks again

**Comment by devnexen** [comment]
Sorry for my poor commits will try to do better if  do a PR anytime.

**Comment by vuvova** [comment]
It turns out that rocksdb check is wrong, it always disables rocksdb unconditionally ☹

For the test to work it needs --std=gnu++11 (or something like that) and -lrt.
I've changed it to
```
-CHECK_CXX_SOURCE_COMPILES(
-"
-#include <time.h>
-int main() {
-  timer_t t;
-  timer_delete(t);
-  return 0;
-}
-" HAVE_TIMER_DELETE)
-
+CHECK_LIBRARY_EXISTS(rt timer_delete "" HAVE_TIMER_DELETE)
```

Now it succeeds on Linux, hope it still fails on OpenBSD, as you want it to.

File: storage/tokudb/CMakeLists.txt (hunk 4 above)
Comment (jmuchemb) [reply storage/tokudb/CMakeLists.txt]: Extending TOKUDB_OK is wrong because these 2 flags are optional features. Our machines don't have F_NOCACHE but we use TokuDB. Now with 10.2.11, MariaDB does not build TokuDB anymore.

Comment (jmuchemb) [reply storage/tokudb/CMakeLists.txt]: Hmm... I should have looked at the bug tracker first. This is already marked as fixed (but I don't know in which branch).
